federation/pkg/kubefed/util/util.go

```diff
@@ -259,15 +261,22 @@ func buildConfigFromSecret(secret *api.Secret, serverAddress string) (*restclien
 // clientset for either the preferred version or the first listed version (if no preference listed)
 // TODO: We need to evaluate the usage of RESTMapper interface to achieve te same functionality
 func GetVersionedClientForRBACOrFail(hostFactory cmdutil.Factory) (client.Interface, error) {
-	discoveryclient, err := hostFactory.DiscoveryClient()
```

Comment:
discoveryclient has `Fresh()` and `Invalidate()` functions that make it easy to tell if cached data was used, and request new data. I'd expect a usage like this:

```go
func GetVersionedClientForRBACOrFail(hostFactory cmdutil.Factory) (client.Interface, error) {
	discoveryclient, err := hostFactory.DiscoveryClient()
	if err != nil {
		return nil, err
	}
	rbacVersion, err := getRBACVersion(discoveryclient)
	if err != nil && !discoveryclient.Fresh() {
		discoveryclient.Invalidate()
		rbacVersion, err = getRBACVersion(discoveryclient)
	}
	if err != nil {
		return nil, err
	}
	return hostFactory.ClientSetForVersion(&gv)
}
func getRBACVersion(discoveryclient discovery.CachedDiscoveryInterface) (*schema.GroupVersion, error) {
	groupList, err := discoveryclient.ServerGroups()
	if err != nil {
		return nil, fmt.Errorf("Couldn't get clientset to create RBAC roles in the host cluster: %v", err)
 	}
	for _, g := range groupList.Groups {
		if g.Name == rbac.GroupName {
			if g.PreferredVersion.GroupVersion != "" {
				gv, err := schema.ParseGroupVersion(g.PreferredVersion.GroupVersion)
 				if err != nil {
 					return nil, err
 				}
 				return &gv, nil
 			}
			for _, version := range g.Versions {
				if version.GroupVersion != "" {
					gv, err := schema.ParseGroupVersion(version.GroupVersion)
					if err != nil {
						return nil, err
					}
					return &gv, nil
				}
			}
		}
	}
	return nil, &NoRBACAPIError{rbacAPINotAvailable}
}
```